src/control/system/database_members.go

```diff
@@ -147,4 +149,5 @@ func (mdb *MemberDatabase) removeMember(m *Member) {
 	delete(mdb.Ranks, m.Rank)
 	delete(mdb.Uuids, m.UUID)
 	delete(mdb.Addrs, m.Addr.String())
+	// TODO KJ: remove fault domain if no one else needs it
```

Comment:
I'd like some feedback on this. I was trying to come up with ways to track usage of a domain, was thinking of incorporating some reference-counting scheme in the tree itself, or maybe maintaining a separate map[string]int with fault domain string -> count. But I'm not sure how important it is to clean up the tree to begin with. How often will a domain actually disappear?

Reply:
Hmm. Members are always going to be leaf nodes, right? Unless I'm missing something, it seems like it should always be safe to delete the leaf domain. I guess the question is whether or not you then walk back up to remove any intermediate nodes that are now dangling because they're not connected to a leaf.

In practice, I don't expect members to be removed very often. I think that typically they'll be replaced rather than flat-out removed.

Reply:
it doesn't seem like you need to be doing much cleanup of the tree at the moment because it's unlikely to be particular dynamic during the lifetime of the database, would think that removing the member/leaf node would suffice for the moment

Reply:
I think in practice the leaves on the MS side will be the hosts. So there could be multiple members using the host domain. Right now the IO servers deal with their own rank/target level domains, so my instinct is not to worry about them in the MS.

Reply:
OK. Not sure if it makes sense, but it occurs to me that if you did make ranks the leaf nodes in this tree, you would have a way to easily tell whether or not the host node still has children.

Reply:
Incidentally, I agree that we probably don't need to be concerned with targets or anything under the rank at this level.

Reply:
Good point. I'll try adding the rank nodes.

Reply:
removeMember() concerns a rank, I thought that's what this discussion was about and that the leaf nodes were the ranks

Reply:
@tanabarr Right, that's what it is about. But the IO server code effectively adds the rank nodes and below, so I was thinking we didn't need to track ranks at this level, instead only tracking the user-supplied domain above the rank. Based on this discussion, though, it does seem like it would be valuable to include the ranks in our tree, even if we don't opt to pass them to the IOserver.